tasm/tiles/src/LayoutDatabase.cc

```diff
@@ -0,0 +1,386 @@
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+    ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+
+    int result;
+    std::vector<unsigned int> widthsOfColumns;
+    while ((result = sqlite3_step(select)) == SQLITE_ROW) {
+        widthsOfColumns.push_back(sqlite3_column_int(select, 0));
+    }
+
+    ASSERT_SQLITE_DONE(result);
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+
+    return widthsOfColumns;
+}
+
+std::vector<unsigned int> LayoutDatabase::heightsForId(const std::string &video, unsigned int id) const {
+    // select heights
+    std::string query = "SELECT height FROM heights WHERE video = ? AND id = ? ORDER BY position ASC";
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+    ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, video.c_str(), -1, SQLITE_STATIC));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+
+    int result;
+    std::vector<unsigned int> heightsOfRows;
+    while ((result = sqlite3_step(select)) == SQLITE_ROW) {
+        heightsOfRows.push_back(sqlite3_column_int(select, 0));
+    }
+
+    ASSERT_SQLITE_DONE(result);
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+
+    return heightsOfRows;
+}
+
+std::experimental::filesystem::path LayoutDatabase::locationOfTileForId(const std::shared_ptr<TiledEntry> entry, unsigned int tileNumber, int id) const {
+    std::string query = "SELECT first_frame, last_frame FROM layouts WHERE video = ? AND id = ?";
+    sqlite3_stmt *select;
+    ASSERT_SQLITE_OK(sqlite3_prepare_v2(db_, query.c_str(), query.length(), &select, nullptr));
+    ASSERT_SQLITE_OK(sqlite3_bind_text(select, 1, entry->name().c_str(), -1, SQLITE_STATIC));
+    ASSERT_SQLITE_OK(sqlite3_bind_int(select, 2, id));
+
+    int result;
+    unsigned int firstFrame;
+    unsigned int lastFrame;
+    if ((result = sqlite3_step(select)) == SQLITE_ROW) {
+        firstFrame = sqlite3_column_int(select, 0);
+        lastFrame = sqlite3_column_int(select, 1);
+    }
+
+    ASSERT_SQLITE_DONE(sqlite3_step(select));
+    ASSERT_SQLITE_OK(sqlite3_finalize(select));
+
+    std::experimental::filesystem::path directory = TileFiles::directoryForTilesInFrames(entry->path(), id, firstFrame, lastFrame);
+
+    return TileFiles::tileFilename(directory, tileNumber);
+}
+
+void LayoutDatabase::addTileLayout(const std::string &video, unsigned int id, unsigned int firstFrame, unsigned int lastFrame, const std::shared_ptr<TileLayout> tileLayout) {
+    addTileBasicLayout(video, id, firstFrame, lastFrame, tileLayout->numberOfColumns(), tileLayout->numberOfRows());
+    addTileLayoutWidths(video, id, tileLayout->widthsOfColumns());
+    addTileLayoutHeights(video, id, tileLayout->heightsOfRows());
```

Comment:
I think we want to do this inside of a single transaction so that if any step fails, we won't end up with partial state. An example of this is in `SemanticIndexSQLiteBase::addBulkMetadata()`.